### PR TITLE
Feat/Opcode Circuits for DIV, REM, and REMU opcodes

### PR DESCRIPTION
**Pull request by bgillesp**
This PR extends the existing `DIVU` opcode implementation to expose the remainder as an output, and to handle the (more complicated) signed case.  See the comments in `div.rs` for details, but in summary:
- The basic design of signed and unsigned `DIV` and `REM` opcode circuits is to represent the division algorithm relation `dividend = quotient * divisor + remainder` using native field elements, and constraining so that either `0 <= remainder < divisor` or `divisor == 0` holds.
- Some nontrivial steps are needed in the signed case to handle the additional complexity of different signs for the divisor and remainder values, as well as to handle signed division overflow where `-2^31 / -1 = 2^31` is out of the representable range for `i32` values.
- In either case, either the `quotient` or the `remainder` value is exposed as the output register `rd`.
- In theory with the current implementation, consecutive calls to `DIV` and `REM` or `DIVU` and `REMU` could be batched into a single circuit since both of the outputs are represented; however, this is not yet implemented.

Currently the circuit implementation and documentation are finished and ready for review.  Unit tests still need to be finished, but it would be great to get some eyes on the circuit design now.

### CODE REVIEW

**Comment by kunxian-xia** [comment]
@bgillesp when will this pr be ready for review?

**Comment by bgillesp** [comment]
Unit tests still need to be added for the new opcode circuits, but everything else is in shape for an initial review -- will change the PR out of Draft now.

File: ceno_zkvm/src/gadgets/is_lt.rs
Comment (matthiasgoergens) [inline ceno_zkvm/src/gadgets/is_lt.rs:21]: See https://github.com/scroll-tech/ceno/pull/731

**Review by matthiasgoergens — COMMENTED** [review]
A few comments and suggestions so far.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (kunxian-xia) [reply ceno_zkvm/src/instructions/riscv/div.rs:1]: #733 is merged.

**Comment by matthiasgoergens** [comment]
@bgillesp I made a version of this PR that has an easier to read diff and left some comments there.  They apply here as well.  See https://github.com/scroll-tech/ceno/pull/734

**Comment by hero78119** [comment]
Attach DIVU circuit statistics,

First row master branch, second row this PR
```
+---------------+---------------+---------+-------+-----------+--------+------------+---------------------+
| opcode_name   | num_instances | lookups | reads | witnesses | writes | 0_expr_deg | 0_expr_sumcheck_deg |
+---------------+---------------+---------+-------+-----------+--------+------------+---------------------+
| DIVU          | 0             | 21      | 4     | 37        | 4      | [1: 7]     | [2: 8]              | 
| DIVU          | 0             | 13      | 4     | 29        | 4      | [1: 5]     | [2: 5]              | 
```

**Comment by matthiasgoergens** [comment]
> Attach DIVU circuit statistics 🔥
> 
> First row master branch, second row this PR
> 
> ```
> +---------------+---------------+---------+-------+-----------+--------+------------+---------------------+
> | opcode_name   | num_instances | lookups | reads | witnesses | writes | 0_expr_deg | 0_expr_sumcheck_deg |
> +---------------+---------------+---------+-------+-----------+--------+------------+---------------------+
> | DIVU          | 0             | 21      | 4     | 37        | 4      | [1: 7]     | [2: 8]              | 
> | DIVU          | 0             | 13      | 4     | 29        | 4      | [1: 5]     | [2: 5]              | 
> ```

Keep in mind that the PR has at least one bug (and that might affect the stats)?

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (KimiWu123) [inline ceno_zkvm/src/instructions/riscv/div.rs:192]: This should be `remainder` as Matt mentioned.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (KimiWu123) [inline ceno_zkvm/src/instructions/riscv/div.rs:364]: You can use `assign_value` instead.

File: ceno_zkvm/src/gadgets/signed.rs
Comment (KimiWu123) [inline ceno_zkvm/src/gadgets/signed.rs:42]: Will it be better to use `i32`? Callers could assign `i32::MAX + 1`, and this could cause unexpeced result If the type is`u32`.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (KimiWu123) [inline ceno_zkvm/src/instructions/riscv/div.rs:296]: The same here, can use `u32::MAX` directly.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (KimiWu123) [inline ceno_zkvm/src/instructions/riscv/div.rs:215]: This curcuit is limited to 32-bits registers and Goldiliock field so I think we can just use `u32::MAX` here.

File: ceno_zkvm/src/gadgets/signed.rs
Comment (hero78119) [inline ceno_zkvm/src/gadgets/signed.rs:32]: could reused util function defined on `uint` to deal with the conversion `to_field_expr(is_negative)`

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (hero78119) [inline ceno_zkvm/src/instructions/riscv/div.rs:192]: `&remainder` This could be captured once unittest added

**Review by hero78119 — COMMENTED** [review]
Epic job! thanks for the implementation and clear documentation.
Just do a quick overview and will proceed arithmetics constrain review 2nd round

File: ceno_zkvm/src/gadgets/signed_ext.rs
Comment (bgillesp) [reply ceno_zkvm/src/gadgets/signed_ext.rs:13]: Sounds good, modified!

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (bgillesp) [reply ceno_zkvm/src/instructions/riscv/div.rs:215]: Yep, that's fair, the limitation is pretty hard-coded throughout, and this is definitely more readable.  Will make the change.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (bgillesp) [reply ceno_zkvm/src/instructions/riscv/div.rs:1]: Finished merging in master, hopefully the diff is a little nicer now.

File: ceno_zkvm/src/gadgets/signed_ext.rs
Comment (bgillesp) [reply ceno_zkvm/src/gadgets/signed_ext.rs:9]: Yep, makes sense -- updated!

**Comment by bgillesp** [comment]
Thank you for the generous feedback everyone!  :pray:  I think I've addressed your comments as much as I can currently -- will keep an eye on this space to help move things forward as the review progresses.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (KimiWu123) [inline ceno_zkvm/src/instructions/riscv/div.rs:202]: Really minor one, I like the original style more because It's clear and slightly efficient. It reduces the number of recursion [here](https://github.com/scroll-tech/ceno/blob/master/ceno_zkvm/src/expression/monomial.rs#L13). But it depends on you :)
```rust                    
// a_neg * (1 - b_neg) + (1 - a_neg) * b_neg
(a_neg.clone() + b_neg.clone()) - (Expression::<E>::from(2) * a_neg * b_neg)
```

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (KimiWu123) [inline ceno_zkvm/src/instructions/riscv/div.rs:469]: I guess you forgot this `TODO`. I didn't see any testing added.

Comment (bgillesp) [reply ceno_zkvm/src/instructions/riscv/div.rs:469]: Not forgotten!  I mentioned that this is still TODO in the PR description -- another member of our team is working on it, but paused yesterday due to the rapid changes from the review.  Should be finished relatively soon, but probably an initial review of the circuit design would still be valuable at this point before we validate with unit tests.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (bgillesp) [reply ceno_zkvm/src/instructions/riscv/div.rs:202]: Yeah, I think the fact that it was simplified to monomial form was why I went with the original version, but then the current revision is closer to the underlying logic of the expression, which gives an easily readable meaning for binary inputs.  I don't think I have a strong opinion about it -- any thoughts @matthiasgoergens since the change was your suggestion?

File: ceno_zkvm/src/gadgets/signed.rs
Comment (bgillesp) [reply ceno_zkvm/src/gadgets/signed.rs]: Just merged this in from master and ported the `i32::from(val)` syntax from the `Signed` implementation in `mul.rs`, so should be all set here.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (matthiasgoergens) [reply ceno_zkvm/src/instructions/riscv/div.rs:202]: There are two parts to my suggestion.  One is to remove the manual clones, and the other is to use a simpler algebraic expression.

They are independent of each other.

I don't really have much of an opinion on whether `(a + b - 2 * a * b)` or `a * (1 - b) + (1 - a) * b` is easier to read.  I went with the latter, because someone had made a comment `// a_neg * (1 - b_neg) + (1 - a_neg) * b_neg` to that end; so I assume someone thought that form is easier to read.

Here's my suggestion, if you want to use the original expression:

```rust
// a_neg ^ b_neg
&a_neg + &b_neg - 2 * a_neg * b_neg
```

Btw, when you say 'efficient', do you mean that the distribution function will be called slightly less often; or do you mean that the final resulting expression will be different?

Comment (bgillesp) [reply ceno_zkvm/src/instructions/riscv/div.rs]: Sure, this seems also like a good way to write it, with the simple comment -- updated to this version in the latest commit.

**Comment by bgillesp** [comment]
Note that while commit `e719763` is showing failed tests, the `cargo make tests` command runs successfully on my local build, so I'm guessing it might be something stalled on the CI end.

**Comment by matthiasgoergens** [comment]
> Note that while commit `e719763` is showing failed tests, the `cargo make tests` command runs successfully on my local build, so I'm guessing it might be something stalled on the CI end.

I clicked the button to kick off a re-run of the tests.  (Is that button available for you?)

File: ceno_zkvm/src/gadgets/is_lt.rs
Comment (matthiasgoergens) [inline ceno_zkvm/src/gadgets/is_lt.rs:21]:
```suggestion
```

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (matthiasgoergens) [inline ceno_zkvm/src/instructions/riscv/div.rs:224]: I suspect we actually might be able to get away without any special handling of signed division overflow (and just get the right answer out of the general constraints).

But that's something for me to think about later, nothing that should hold up this PR.

> My suspicion is because we have more range available in Goldilocks, and (i32::MIN as Goldilocks) / (-1) == 1<<31 as Goldilocks, which is the answer we want at the end.

**Review by matthiasgoergens — APPROVED** [review]
Thanks for your patience with the review process.  Looking good now.

Please remove the now obsolete TODO about the LT -> Lt rename before merging.

**Comment by matthiasgoergens** [comment]
Is there anything left to do here, or can we merge? @bgillesp Do you have enough rights to the repository to hit the 'Merge when ready' button yourself, or do you need someone else to do it for you?

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (hero78119) [inline ceno_zkvm/src/instructions/riscv/div.rs:60]: This example need fix

**Review by hero78119 — COMMENTED** [review]
some review feedback of docs :)

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (hero78119) [inline ceno_zkvm/src/instructions/riscv/div.rs:209]: is_dividend_`min`_negative to match with i32::MIN syntatically

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (hero78119) [inline ceno_zkvm/src/instructions/riscv/div.rs:215]: not sure syntax, but seems `is_divisor_neg_one` fit better

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (hero78119) [reply ceno_zkvm/src/instructions/riscv/div.rs:224]: I think the key point here is not Goldilocks field range, it's riscv signed 32 bit registet can not represent overflow 2^31. So riscv define a special case for it.

**Review by hero78119 — COMMENTED** [review]
reviewed on the arithmetics & constrains design. Docs explanations fit well with constrains design. 👍 
Stay tune for for unittest added then it's good to go!

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (matthiasgoergens) [reply ceno_zkvm/src/instructions/riscv/div.rs:224]: I know.  I am saying that the special case disappears in our setting!

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (matthiasgoergens) [reply ceno_zkvm/src/instructions/riscv/div.rs:469]: I tried to write a simple test, but it kept failing.  (Not sure that's because I made some mistakes, or because the code here is wrong.)

So I would suggest that we only merge this PR once we have some passing tests.

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (matthiasgoergens) [inline ceno_zkvm/src/instructions/riscv/div.rs:182]: We can do that rename in its own PR, to make this one easier to review:

https://github.com/scroll-tech/ceno/pull/766

File: ceno_zkvm/src/instructions/riscv/div.rs
Comment (bgillesp) [reply ceno_zkvm/src/instructions/riscv/div.rs:190]: Agreed the "max negative" phrasing is a bit awkward -- is `is_dividend_signed_min` okay?

**Comment by bgillesp** [comment]
Just incorporated the updated interface for signed instance assignment to the `AssertLtConfig` gadget used by the `remainder_nonnegative` field, fixed up the documentation to address the bug we found in normalizing the signed remainder, and addressed some comments from @hero78119.  The new tests are passing with the updated assignment, and I think this should be good to go.  Give it one last look over @mcalancea?

**Review by mcalancea — COMMENTED** [review]
Unit tests added, releasing my requested changes.

**Review by hero78119 — APPROVED** [review]
👍 LGTM!

**Review by mcalancea — APPROVED** [review]

**Pull request merged**
